### PR TITLE
feat: extract logic in 'submitRequest'

### PR DESCRIPTION
**Pull request by k-g-j**
## Description

This PR extracts the `#submitRequest` logic into smaller helper functions and utilities.
Functions added:
* `#validateMethod`
* `#createPromise`
* `#submitSnapRequest`
* `#handleSyncResponse`
* `#handleAsyncResponse`
* `#validateRedirectUrl`

Addresses: [Comment point 3](https://github.com/MetaMask/eth-snap-keyring/pull/228#discussion_r1506770301) on #228 

## Testing

All current tests run as expected, and coverage remains at 100%

### CODE REVIEW

File: src/SnapKeyring.ts [hunk 1]
```diff
@@ -397,68 +398,173 @@ export class SnapKeyring extends EventEmitter {
     chainId?: string;
   }): Promise<Json> {
     const { account, snapId } = this.#resolveAddress(address);
-    if (!account.methods.includes(method as EthMethod)) {
-      throw new Error(
-        `Method '${method}' not supported for account ${account.address}`,
-      );
-    }
+    this.#validateMethod(method as EthMethod, account);
```
Comment (ccharly) [inline src/SnapKeyring.ts:401]: I would go for `hasMethod` instead, and maybe have the `account` as a first argument (more like if we would use `account` as an object/receiver for the method).

This is purely cosmetic, so you can disagree 🙈

File: src/SnapKeyring.ts [hunk 2]
```diff
@@ -397,68 +398,173 @@ export class SnapKeyring extends EventEmitter {
...
+
     const requestId = uuid();
 
     // Create the promise before calling the snap to prevent a race condition
     // where the snap responds before we have a chance to create it.
-    const promise = new DeferredPromise<Response>();
-    this.#requests.set(requestId, { promise, snapId });
+    const promise = this.#createPromise<Response>(requestId, snapId);
```
Comment (ccharly) [inline src/SnapKeyring.ts:407]: `createPromise` seems a bit too generic IMO, maybe `createRequestPromise` (as we're also registering it to the request map)

File: src/SnapKeyring.ts (hunk 1 above)
Comment (k-g-j) [reply src/SnapKeyring.ts]: I like it 😄
Implemented in this commit [0522eb3](https://github.com/MetaMask/eth-snap-keyring/pull/232/commits/0522eb325e4abf8bfc279b1891737872fda20f3a)

File: src/SnapKeyring.ts (hunk 2 above)
Comment (k-g-j) [reply src/SnapKeyring.ts]: Nice point! Here you go [e29f479](https://github.com/MetaMask/eth-snap-keyring/pull/232/commits/e29f479809383fa56d6be2abb7f08c034670546a) ☺️

**Review by danroc — APPROVED** [review]

**Review by danroc — DISMISSED** [review]

File: src/SnapKeyring.ts [hunk 3]
```diff
@@ -397,68 +398,173 @@ export class SnapKeyring extends EventEmitter {
...
+    requestId: string,
+    snapId: SnapId,
+  ): DeferredPromise<Response> {
+    const promise = new DeferredPromise<Response>();
+    this.#requests.set(requestId, { promise, snapId });
+    return promise;
+  }
+
+  /**
+   * Submits a request to a Snap and handles the response.
```
Comment (ccharly) [inline src/SnapKeyring.ts:464]: Strictly speaking, we do not handles the response, as this is done by the caller with the respective `{Sync,Async}Response` helpers.

Unless you were talking about the error cases here?

File: src/SnapKeyring.ts [hunk 4]
```diff
@@ -397,68 +398,173 @@ export class SnapKeyring extends EventEmitter {
...
+   *
+   * @param options - The options for the snap request.
```
Comment (ccharly) [inline src/SnapKeyring.ts:466]: Really really minor thing, we sometimes use `snap` and sometimes `Snap`, IDK which one is best, but we could "align" most comments to one notation, WDYT?

Comment (danroc) [reply src/SnapKeyring.ts:466]: `Snap` if the preferred form, AFAIK

Comment (k-g-j) [reply src/SnapKeyring.ts]: done ✅ [88c3b16](https://github.com/MetaMask/eth-snap-keyring/pull/232/commits/88c3b16e7d3d127d5c824d3fc78fd10f2d74c6f6)

**Review by ccharly — APPROVED** [review]
Awesome!

**Pull request merged**